File: ANSv4/Catalogo3/classes/InterDeposito.vb
Option Explicit On

Public Class InterDeposito

    ' Define como se llama este modulo para el control de errores
    Private Const m_sMODULENAME_ As String = "clsInterDeposito"

    Private Conexion1 As System.Data.OleDb.OleDbConnection  'diego     Private Conexion1 As OleDbConnection

    Private mvarNroInterDeposito As String
    Private mvarBco_Dep_Tipo As String
    Private mvarBco_Dep_Fecha As Date
    Private mvarBco_Dep_Numero As Long
    Private mvarBco_Dep_Monto As Single
    Private mvarBco_Dep_Ch_Cantidad As Byte
    Private mvarBco_Dep_IdCta As Byte
    Private mvarIdCliente As Long

    Private IntDepFacturas As Collection

    Public Event GuardoOK(ByVal NroInterDeposito As String)

    Public Property NroInterDeposito() As String
        Set(ByVal value As String)
            mvarNroInterDeposito = value
        End Set
        Get
            Return mvarNroInterDeposito
        End Get
    End Property

    Public Property Bco_Dep_Tipo() As String

        Set(ByVal value As String)
            mvarBco_Dep_Tipo = value
        End Set
        Get
            Return mvarBco_Dep_Tipo
        End Get
    End Property

    Public Property Bco_Dep_Fecha() As Date
        Set(ByVal value As Date)
            mvarBco_Dep_Fecha = value
        End Set
        Get
            Return mvarBco_Dep_Fecha
        End Get
    End Property

    Public Property Bco_Dep_Numero() As Long
        Set(ByVal value As Long)
            mvarBco_Dep_Numero = value
        End Set
        Get
            Return mvarBco_Dep_Numero
        End Get
    End Property

    Public Property Bco_Dep_Monto() As Single
        Set(ByVal value As Single)
            mvarBco_Dep_Monto = value
        End Set

        Get
            Return mvarBco_Dep_Monto
        End Get
    End Property


    Public Property Bco_Dep_Ch_Cantidad() As Byte
        Set(ByVal value As Byte)
            mvarBco_Dep_Ch_Cantidad = value
        End Set
        Get
            Return mvarBco_Dep_Ch_Cantidad
        End Get
    End Property


    Public Property Bco_Dep_IdCta() As Byte
        Set(ByVal value As Byte)
            mvarBco_Dep_IdCta = value
        End Set

        Get
            Return mvarBco_Dep_IdCta
        End Get
    End Property

    Public ReadOnly Property IdCliente() As Long
        Get
            Return mvarIdCliente
        End Get
    End Property

    Public ReadOnly Property CantidadItems() As Long
        Get
            Return IntDepFacturas.Count
        End Get
    End Property

    'diego    Private Sub Class_Terminate()
    'diego         IntDepFacturas = Nothing
    'diego     End Sub

    Public Sub Nuevo(Optional ByVal IDdeCliente As Long = 0)

        IntDepFacturas = Nothing
        IntDepFacturas = New Collection
        mvarNroInterDeposito = 0

        If IDdeCliente > 0 Then mvarIdCliente = IDdeCliente

    End Sub


    Public Function Renglon(ByVal Numero As Integer) As InterDepositoFacturas
        Return IntDepFacturas("_" & Numero)
    End Function

    Public WriteOnly Property Conexion() As System.Data.OleDb.OleDbConnection
        Set(ByVal value As System.Data.OleDb.OleDbConnection)
            Conexion1 = value
        End Set
    End Property

    Private Function ValidarConexion() As Boolean
        Return Not (Conexion1 Is Nothing)
    End Function

    Public Sub Guardar(Optional ByVal Origen As String = vbNullString)

        If Not (ValidarConexion()) Then Exit Sub

        On Error GoTo ErrorHandler

        Dim rec As System.Data.OleDb.OleDbDataReader
        Dim adoCMD As New System.Data.OleDb.OleDbCommand

        If UCase(Origen) = "VER" Then
            Funciones.adoModulo.adoComandoIU(Conexion1, "DELETE FROM tblInterDeposito WHERE NroInterDeposito='09999-99999999'")
            mvarNroInterDeposito = "09999-99999999"
        Else
            rec = Funciones.adoModulo.adoComando(Conexion1, "SELECT TOP 1 right(NroInterdeposito,8) AS NroInterDeposito FROM tblInterDeposito WHERE left(NroInterDeposito,5)=" & vg.NroUsuario & " ORDER BY NroInterDeposito DESC")
            If Not rec.HasRows Then
                If vg.miSABOR = 2 Then ' ES UN CLIENTE
                    mvarNroInterDeposito = Trim(vg.NroUsuario) & "-C0000001"
                Else
                    mvarNroInterDeposito = Trim(vg.NroUsuario) & "-00000001"
                End If
            Else
                If vg.miSABOR = 2 Then ' ES UN CLIENTE
                    mvarNroInterDeposito = Trim(vg.NroUsuario) & "-C" & Format(CLng(Right(rec("NroInterDeposito"), 7)) + 1, "0000000")
                Else
                    mvarNroInterDeposito = Trim(vg.NroUsuario) & "-" & Format(CLng(Right(rec("NroInterDeposito"), 7)) + 1, "00000000")
                End If
            End If
            rec = Nothing

        End If

        adoCMD.Parameters.Add("pNroInterDeposito", SqlDbType.VarChar, 14).Value = mvarNroInterDeposito
        adoCMD.Parameters.Add("pIdCliente", SqlDbType.Int).Value = mvarIdCliente
        adoCMD.Parameters.Add("pBco_Dep_Tipo", SqlDbType.VarChar, 1).Value = mvarBco_Dep_Tipo
        adoCMD.Parameters.Add("pBco_Dep_Fecha", SqlDbType.Date).Value = mvarBco_Dep_Fecha
        adoCMD.Parameters.Add("pBco_Dep_Numero", SqlDbType.Int).Value = mvarBco_Dep_Numero
        adoCMD.Parameters.Add("pBco_Dep_Monto", SqlDbType.Float).Value = mvarBco_Dep_Monto
        adoCMD.Parameters.Add("pBco_Dep_Ch_Cantidad", SqlDbType.Int).Value = mvarBco_Dep_Ch_Cantidad
        adoCMD.Parameters.Add("pBco_Dep_IdCta", SqlDbType.TinyInt).Value = mvarBco_Dep_IdCta

        adoCMD.Connection = Conexion1
        adoCMD.CommandType = CommandType.StoredProcedure
        adoCMD.CommandText = "usp_InterDeposito_add"
        adoCMD.ExecuteNonQuery()


        adoCMD = Nothing

        Dim I As Integer

        For I = 1 To IntDepFacturas.Count
            GuardarItem(Renglon(I).T_Comprobante, Renglon(I).N_Comprobante, CSng(Renglon(I).Importe))
        Next I

        vg.NroImprimir = mvarNroInterDeposito
        If UCase(Origen) = "VER" Then
            ' Actualizo Fecha de Transmicion para que no lo mande
            Funciones.adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Pedido_Transmicion_Upd '" & mvarNroInterDeposito & "'")
        Else
            Nuevo()
            RaiseEvent GuardoOK(vg.NroImprimir)
        End If

        Exit Sub

ErrorHandler:

        rec = Nothing
        adoCMD = Nothing
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Sub

    Public Sub ADDFacturas(ByVal T_Comprobante As String, ByVal N_Comprobante As String, ByVal Importe As Single)

        Dim mvarItem As New InterDepositoFacturas

        mvarItem.T_Comprobante = T_Comprobante
        mvarItem.N_Comprobante = N_Comprobante
        mvarItem.Importe = Importe

        IntDepFacturas.Add(mvarItem, "_" & IntDepFacturas.Count + 1)

        mvarItem = Nothing

    End Sub

    Private Sub GuardarItem(ByVal pT_Comprobante As String, ByVal pN_Comprobante As String, ByVal pImporte As Single)

        If Not (ValidarConexion()) Then Exit Sub

        On Error GoTo ErrorHandler

        Dim cmd As New System.Data.OleDb.OleDbCommand

        cmd.Parameters.Add("pNroInterDeposito", SqlDbType.VarChar, 14).Value = mvarNroInterDeposito
        cmd.Parameters.Add("pT_Comprobante", SqlDbType.VarChar, 3).Value = pT_Comprobante
        cmd.Parameters.Add("pN_Comprobante", SqlDbType.VarChar, 12).Value = pN_Comprobante
        cmd.Parameters.Add("pImporte", SqlDbType.Float).Value = pImporte

        cmd.Connection = Conexion1
        cmd.CommandType = CommandType.StoredProcedure
        cmd.CommandText = "usp_InterDeposito_Fac_add"
        cmd.ExecuteNonQuery()

        cmd = Nothing
        Exit Sub

ErrorHandler:

        cmd = Nothing
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Sub

End Class

File: ANSv4/Catalogo4/_interdeposito/InterDeposito.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Catalogo._interdeposito
{
    public class InterDeposito
    {
        private const string m_sMODULENAME_ = "clsInterDeposito";

        private System.Data.OleDb.OleDbConnection mvarConexion;
        private string mvarNroInterDeposito;
        private string mvarBco_Dep_Tipo;
        private System.DateTime mvarBco_Dep_Fecha;
        private int mvarBco_Dep_Numero;
        private float mvarBco_Dep_Monto;
        private byte mvarBco_Dep_Ch_Cantidad;
        private byte mvarBco_Dep_IdCta;

        private int mvarIdCliente;
        private byte mvarNroImpresion;
        private string mvarObservaciones;

        private System.Collections.Generic.Dictionary<string, InterDepositoFacturas> IntDepFacturas;

        public event GuardoOKEventHandler GuardoOK;
        public delegate void GuardoOKEventHandler(string NroInterDeposito);

        public InterDeposito(System.Data.OleDb.OleDbConnection conexion, string NroUsuario, int IDDeCliente)
        {
            Nuevo(conexion, IDDeCliente);
        }

       protected void Nuevo(System.Data.OleDb.OleDbConnection conexion, int IDdeCliente = 0)
       {
           IntDepFacturas = new Dictionary<string, InterDepositoFacturas>();
           mvarNroInterDeposito = "";
           mvarConexion = conexion;

           if (IDdeCliente > 0)
               mvarIdCliente = IDdeCliente;
       }

        public string Observaciones
        {
            get { return mvarObservaciones; }
            set { mvarObservaciones = value; }
        }

        public  int  CantidadItems
        {
            get { return IntDepFacturas.Count; }
        }

        // FUNDAMENTAL PARA QUE TE DE LOS NOMBRES
        public InterDepositoFacturas Renglon(int Numero)
        {
            return IntDepFacturas["_" + Numero];
        }

        public System.Data.OleDb.OleDbConnection Conexion
        {
            set { mvarConexion = value; }
        }

        public byte NroImpresion
        {
            get { return mvarNroImpresion; }
            set { mvarNroImpresion = value; }
        }

        public string NroInterDeposito
        {
            get { return mvarNroInterDeposito; }
            set { mvarNroInterDeposito = value; }
        }

        public string Bco_Dep_Tipo
        {
            get { return mvarBco_Dep_Tipo; }
            set { mvarBco_Dep_Tipo = value; }
        }

        public System.DateTime Bco_Dep_Fecha
        {
            get { return mvarBco_Dep_Fecha; }
            set { mvarBco_Dep_Fecha = value; }
        }

        public int Bco_Dep_Numero
        {
            get { return mvarBco_Dep_Numero; }
            set { mvarBco_Dep_Numero = value; }
        }

        public float Bco_Dep_Monto
        {
            get { return mvarBco_Dep_Monto; }
            set { mvarBco_Dep_Monto = value; }
        }


        public byte Bco_Dep_Ch_Cantidad
        {
            get { return mvarBco_Dep_Ch_Cantidad; }
            set { mvarBco_Dep_Ch_Cantidad = value; }
        }

        public byte Bco_Dep_IdCta
        {
            get { return mvarBco_Dep_IdCta; }
            set { mvarBco_Dep_IdCta = value; }
        }

        private bool ValidarConexion()
        {
            return (mvarConexion != null);
        }

        public void Guardar(string Origen)
        {
            if (!(ValidarConexion()))
                return;

            System.Data.OleDb.OleDbDataReader rec = null;
            System.Data.OleDb.OleDbCommand adoCMD = new System.Data.OleDb.OleDbCommand();

            if (IntDepFacturas.Count < 1)
                return;

            if (Origen.ToUpper() == "VER")
            {
                Funciones.oleDbFunciones.ComandoIU(mvarConexion, "DELETE FROM tblInterDeposito WHERE NroInterDeposito='09999-99999999'");
                mvarNroInterDeposito = "09999-99999999";
            }
            else
            {
                rec = Funciones.oleDbFunciones.Comando(mvarConexion, "SELECT TOP 1 right(NroInterdeposito,8) AS NroInterdeposito FROM tblInterDeposito WHERE left(NroInterdeposito,5)=" + Global01.NroUsuario + " ORDER BY NroInterdeposito DESC");

                if (!rec.HasRows)
                {
                    // ES UN CLIENTE
                    if (Global01.miSABOR == Global01.TiposDeCatalogo.Cliente)
                    {
                        mvarNroInterDeposito = Global01.NroUsuario.Trim() + "-C0000001";
                    }
                    else
                    {
                        mvarNroInterDeposito = Global01.NroUsuario.Trim() + "-00000001";
                    }
                }
                else
                {
                    rec.Read();
                    // ES UN CLIENTE
                    if (Global01.miSABOR == Global01.TiposDeCatalogo.Cliente)
                    {
                        mvarNroInterDeposito = Global01.NroUsuario.Trim() + "-C" + (int.Parse(rec["NroInterDeposito"].ToString().Substring(rec["NroInterDeposito"].ToString().Length - 7)) + 1).ToString().PadLeft(7, '0');
                    }
                    else
                    {
                        mvarNroInterDeposito = Global01.NroUsuario.Trim() + "-" + (int.Parse(rec["NroInterDeposito"].ToString().Substring(rec["NroInterDeposito"].ToString().Length - 8)) + 1).ToString().PadLeft(8, '0');
                    }                  
                }
                rec = null;
            }

            adoCMD.Parameters.Add("pNroInterDeposito", System.Data.OleDb.OleDbType.VarChar, 14).Value = mvarNroInterDeposito;
            adoCMD.Parameters.Add("pIdCliente", System.Data.OleDb.OleDbType.Integer).Value = mvarIdCliente;
            adoCMD.Parameters.Add("pBco_Dep_Tipo", System.Data.OleDb.OleDbType.VarChar, 1).Value = mvarBco_Dep_Tipo;
            adoCMD.Parameters.Add("pBco_Dep_Fecha", System.Data.OleDb.OleDbType.Date).Value = mvarBco_Dep_Fecha;
            adoCMD.Parameters.Add("pBco_Dep_Numero", System.Data.OleDb.OleDbType.Integer).Value = mvarBco_Dep_Numero;
            adoCMD.Parameters.Add("pBco_Dep_Monto", System.Data.OleDb.OleDbType.Single).Value = mvarBco_Dep_Monto;
            adoCMD.Parameters.Add("pBco_Dep_Ch_Cantidad", System.Data.OleDb.OleDbType.TinyInt).Value = mvarBco_Dep_Ch_Cantidad;
            adoCMD.Parameters.Add("pBco_Dep_IdCta", System.Data.OleDb.OleDbType.TinyInt).Value = mvarBco_Dep_IdCta;

            adoCMD.Connection = mvarConexion;
            adoCMD.CommandType = System.Data.CommandType.StoredProcedure;
            adoCMD.CommandText = "usp_InterDeposito_add";

            try
            {
                adoCMD.ExecuteNonQuery();
            }
            catch (System.Data.OleDb.OleDbException ex)
            {
                util.errorHandling.ErrorLogger.LogMessage(ex);
                switch (ex.ErrorCode)
                {
                    case -2147467259:
                        throw new util.errorHandling.RegistroDuplicadoException(ex);
                        //throw (new Exception(ex.Message));   //util.errorHandling.ErrorForm.show();
                }
            }
            catch (Exception ex)
            {
                util.errorHandling.ErrorLogger.LogMessage(ex);
                throw (new Exception(ex.Message));  //util.errorHandling.ErrorForm.show();
            }

            adoCMD = null;

            // Ahora GUARDO los ITEMS de ESTE Devolucion

            int I = 0;

            for (I = 1; I <= IntDepFacturas.Count; I++)
            {
                GuardarItem(Renglon(I).T_Comprobante, Renglon(I).N_Comprobante, Renglon(I).Importe);
            }

            Global01.NroImprimir = mvarNroInterDeposito;

            // Por POLITICA NUESTRA  -- >  SE BORRA
            if (Origen.ToUpper() == "VER")
            {
                // Actualizo Fecha de Transmicion para que no lo mande
                Funciones.oleDbFunciones.ComandoIU(mvarConexion, "EXEC usp_InterDeposito_Transmicion_Upd '" + mvarNroInterDeposito + "'");
            }
            else
            {
                auditoria.Auditor.instance.guardar(auditoria.Auditor.ObjetosAuditados.InterDeposito, auditoria.Auditor.AccionesAuditadas.EXITOSO, "cli:" + mvarIdCliente.ToString().Trim().PadLeft(6, '0') + " intD:" + mvarNroInterDeposito + " tot:");
                //Nuevo();
                if (GuardoOK != null)
                {
                    GuardoOK(Global01.NroImprimir);
                }
            }
        }

        public void ADDFacturas(string T_Comprobante, string N_Comprobante, float Importe)
	    {
		    InterDepositoFacturas mvarItem = new InterDepositoFacturas();

		    mvarItem.T_Comprobante = T_Comprobante;
		    mvarItem.N_Comprobante = N_Comprobante;
		    mvarItem.Importe = Importe;

            IntDepFacturas["_" + (IntDepFacturas.Count + 1).ToString()] = mvarItem;

		    mvarItem = null;
	    }

        private void GuardarItem(string pT_Comprobante, string pN_Comprobante, float pImporte)
        {
            if (!(ValidarConexion()))
                return;

            System.Data.OleDb.OleDbCommand cmd = new System.Data.OleDb.OleDbCommand();

            cmd.Parameters.Add("pNroInterDeposito", System.Data.OleDb.OleDbType.VarChar, 14).Value = mvarNroInterDeposito;
            cmd.Parameters.Add("pT_Comprobante", System.Data.OleDb.OleDbType.VarChar, 3).Value = pT_Comprobante;
            cmd.Parameters.Add("pN_Comprobante", System.Data.OleDb.OleDbType.VarChar, 12).Value = pN_Comprobante;
            cmd.Parameters.Add("pImporte", System.Data.OleDb.OleDbType.Single).Value = pImporte;

            cmd.Connection = mvarConexion;
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            cmd.CommandText = "usp_InterDeposito_Fac_add";
            cmd.ExecuteNonQuery();

            cmd = null;
        }
   
    }
}